src/gdbserver/gdb-server.c

```diff
@@ -38,7 +38,7 @@
 static stlink_t *connected_stlink = NULL;
 static bool semihosting = false;
 static bool serial_specified = false;
-static char serialnumber[28] = {0};
+static uint8_t serialnumber[STLINK_SERIAL_SIZE] = {0}; // XXX? Why was size 28?
```

Comment:
I think the AN you reference is unrelated to serial numbers from the programmers as this is a note for manufacturing serial number programming. Probably the programmers use the STM32 96-bit unique ID as serial number (12 bytes binary, 24 bytes hex encoded).

I prefer to have all serial number code centralized in a few functions and use it in the different tools. Currently this is a mess as you have seen.